3rdParty/fuerte/src/vst.cpp

```diff
@@ -340,77 +337,83 @@ namespace parser {
 
-// isChunkComplete returns the length of the chunk that starts at the given
-// begin
-// if the entire chunk is available.
-// Otherwise 0 is returned.
-std::size_t isChunkComplete(uint8_t const* const begin,
-                            std::size_t const lengthAvailable) {
-  // there is not enought to read the length of
-  if (lengthAvailable < sizeof(uint32_t)) { // first header field
-    return 0;
+/// @brief readChunkHeaderVST1_0 reads a chunk header in VST1.0 format.
+/// @return true if chunk is complete
+bool readChunkHeaderVST1_0(Chunk& chunk, uint8_t const* hdr, std::size_t avail) {
+  if (avail < minChunkHeaderSize) {
+    return false;
   }
-  // read chunk length
-  uint32_t lengthThisChunk = basics::uintFromPersistentLittleEndian<uint32_t>(begin);
-  if (lengthAvailable < lengthThisChunk) {
-    FUERTE_LOG_VSTCHUNKTRACE << "\nchunk incomplete: " << lengthAvailable << "/"
-                             << lengthThisChunk << "(available/len)\n";
-    return 0;
-  }
-  FUERTE_LOG_VSTCHUNKTRACE << "\nchunk complete: " << lengthThisChunk
-                           << " bytes\n";
-  return lengthThisChunk;
-}
-
-// readChunkHeaderVST1_0 reads a chunk header in VST1.0 format.
-Chunk readChunkHeaderVST1_0(uint8_t const* bufferBegin) {
-  Chunk chunk;
-
-  uint8_t const* hdr = bufferBegin;
+  
   chunk.header._chunkLength = basics::uintFromPersistentLittleEndian<uint32_t>(hdr + 0);
   chunk.header._chunkX = basics::uintFromPersistentLittleEndian<uint32_t>(hdr + 4);
   chunk.header._messageID = basics::uintFromPersistentLittleEndian<uint64_t>(hdr + 8);
+  
+  if (avail < chunk.header._chunkLength) {
+    return false;
+  }
+  
   size_t hdrLen = minChunkHeaderSize;
-
-  if (chunk.header.isFirst() && chunk.header.numberOfChunks() > 1) {
-    // First chunk, numberOfChunks>1 -> read messageLength
-    chunk.header._messageLength =
-        basics::uintFromPersistentLittleEndian<uint64_t>(hdr + 16);
-    hdrLen = maxChunkHeaderSize; // first chunk header is bigger
+  if (chunk.header.isFirst()) {
+    if (chunk.header.numberOfChunks() > 1) {
+      if (avail < maxChunkHeaderSize) {
+        return false;
+      }
+      hdrLen = maxChunkHeaderSize; // first chunk header is bigger
+      // First chunk, numberOfChunks>1 -> read messageLength
+      chunk.header._messageLength =
+      basics::uintFromPersistentLittleEndian<uint64_t>(hdr + 16);
+    } else {
+      chunk.header._messageLength = chunk.header._chunkLength - hdrLen;
+    }
+  } else { // not needed / known otherwise
+    chunk.header._messageLength = 0;
   }
 
   size_t contentLength = 0;
   if (chunk.header._chunkLength >= hdrLen) { // prevent underflow
-    chunk.header._chunkLength = chunk.header._chunkLength - hdrLen;
+    contentLength = chunk.header._chunkLength - hdrLen;
   } else {
     FUERTE_LOG_ERROR << "received invalid chunk length";
   }
+  
+  if (avail < contentLength) { // chunk incomplete
```

Comment:
Shouldn't this be `if (avail < contentLength + hdrLen)`?
Because `avail` seems to contain the total number of bytes, including the header, but `contentLength` is the net content length, without the header size.